// succinct/es/reaction/index.js
import Context from '../context';
var REACTION_ID = 'reaction';
var curId = 0;
var Reaction = (function () {
    function Reaction(host) {
        this.host = host;
        this.succinctList = [];
        this.succinctMap = {};
        this.id = REACTION_ID + (curId++);
    }
    Reaction.prototype.track = function (fn) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        Context.enter(this);
        fn.apply(void 0, args);
        Context.leave();
    };
    Reaction.prototype.dispose = function () {
        var _this = this;
        this.succinctMap = null;
        this.succinctList.forEach(function (succinct) {
            succinct && succinct.removeSubscribe(_this);
        });
    };
    Reaction.prototype.addSuccinct = function (succinct) {
        if (this.succinctMap[succinct.id])
            return;
        this.succinctMap[succinct.id] = true;
        this.succinctList.push(succinct);
    };
    return Reaction;
}());
export default Reaction;
